ESCPOS_NET/Printers/BasePrinter.cs

```diff
@@ -71,59 +91,104 @@ public virtual void Write(byte[] bytes)
             }
-            _readBuffer = new ConcurrentQueue<byte>();
-            _monitoring = true;
         }
 
         public virtual void StopMonitoring()
         {
-            _monitoring = false;
-            _readBuffer = new ConcurrentQueue<byte>();
+            if (_isMonitoring)
+            {
+                Debug.WriteLine(nameof(StopMonitoring));
+				_isMonitoring = false;
+				ReadBuffer = new ConcurrentQueue<byte>();
+
+                if (_cancellationTokenSource != null)
+                {
+                    _cancellationTokenSource.Cancel();
+                }                    
+            }
         }
 
-        public virtual void DataAvailable()
+        public virtual void DataAvailable(bool timeout)
         {
-            if (_readBuffer.Count() % 4 == 0)
+            if (timeout)
+            {
+                TryUpdatePrinterStatus(null, timeout);
+                return;
+            }
+
+            if (ReadBuffer.Count() % 4 == 0)
             {
                 var bytes = new byte[4];
                 for (int i = 0; i < 4; i++)
                 {
-                    if (!_readBuffer.TryDequeue(out bytes[i])) return; // Ran out of bytes unexpectedly.
+                    if (!ReadBuffer.TryDequeue(out bytes[i]))
+                    {
+                        // Ran out of bytes unexpectedly.
+                        return; 
+                    }
                 }
 
-                TryUpdatePrinterStatus(bytes);
+                TryUpdatePrinterStatus(bytes, timeout);
+
                 // TODO: call other update handlers.
             }
         }
 
-        private bool _monitoring = false;
-
-        private void TryUpdatePrinterStatus(byte[] bytes)
+        private void TryUpdatePrinterStatus(byte[] bytes, bool timeout)
         {
+            if (timeout)
+            {
+                Status = new PrinterStatusEventArgs()
+                {
+                    DeviceConnectionTimeout = true,
+                };
+
+                StatusChanged?.Invoke(this, Status);
+                return;
+            }
+#if DEBUG
+            var bytesToString = string.Empty;
+            var index = 0;
+            foreach (var b in bytes)
+            {
+                bytesToString += $"index[{index}], value[{b}]\n";
+                index++;
+            }
+
+            Debug.WriteLine($"TryUpdatePrinterStatus: \n{bytesToString}");
+#endif
+
             // Check header bits 0, 1 and 7 are 0, and 4 is 1
             if (bytes[0].IsBitNotSet(0) && bytes[0].IsBitNotSet(1) && bytes[0].IsBitSet(4) && bytes[0].IsBitNotSet(7))
             {
-                Status = new PrinterStatus()
+                Status = new PrinterStatusEventArgs()
                 {
-                    IsCashDrawerOpen = bytes[0].IsBitNotSet(2), // TODO: verify that this state is correct and is not inverted.
```

Comment:
I saw this comment was removed, do you happen to have a cash drawer and know that this is correct?  I don't have a cash drawer so I never checked it.

Reply:
I've tested it on a couple of cash drawer and seems to be correct. I think this could a bit fickle however depending on the different types of cash drawer and if the pins are reset correctly. The Window.PointOfSale libs get this wrong at times too.